diff --git a/apps/nccl/src/allreduce.hpp b/apps/nccl/src/allreduce.hpp
index 7fa7b4e..865e744 100644
--- a/apps/nccl/src/allreduce.hpp
+++ b/apps/nccl/src/allreduce.hpp
@@ -384,25 +384,24 @@ __global__ void __launch_bounds__(1024, 1)
 template <typename T>
 __global__ void __launch_bounds__(512, 1)
     allreduce8Mod(T* buff, T* scratch, T* resultBuff, mscclpp::DeviceHandle<mscclpp::SmChannel>* smChannels,
-               size_t channelScratchOffset, int rank, int nRanksPerNode, int worldSize, size_t nelems) {
+                  size_t channelScratchOffset, int rank, int nRanksPerNode, int worldSize, size_t nelems) {
   const int nPeer = nRanksPerNode - 1;
   const size_t chanOffset = nPeer * blockIdx.x;
   // assume (nelems * sizeof(T)) is divisible by (16 * worldSize)
   const size_t nInt4 = nelems * sizeof(T) / sizeof(int4);
   size_t nInt4PerRank = nInt4 / worldSize;
-  if (nInt4 % worldSize)
-        nInt4PerRank = nInt4PerRank + 1;
+  if (nInt4 % worldSize) nInt4PerRank = nInt4PerRank + 1;
 
   auto smChans = smChannels + chanOffset;
 
-  size_t channelScratchResultOffset = channelScratchOffset + SCRATCH_SIZE/2;
+  size_t channelScratchResultOffset = channelScratchOffset + SCRATCH_SIZE / 2;
 
   int4* buff4 = reinterpret_cast<int4*>(buff);
   int4* scratch4 = reinterpret_cast<int4*>((char*)scratch + channelScratchOffset);
   int4* scratch4Result = reinterpret_cast<int4*>((char*)scratch + channelScratchResultOffset);
   int4* resultBuff4 = reinterpret_cast<int4*>(resultBuff);
 
-      // Distribute `nInt4PerRank` across all blocks with the unit size `unitNInt4`
+  // Distribute `nInt4PerRank` across all blocks with the unit size `unitNInt4`
   constexpr size_t unitNInt4 = 512;
   const size_t maxNInt4PerBlock =
       (((nInt4PerRank + gridDim.x - 1) / gridDim.x) + unitNInt4 - 1) / unitNInt4 * unitNInt4;
@@ -426,7 +425,7 @@ __global__ void __launch_bounds__(512, 1)
 
   __shared__ mscclpp::DeviceHandle<mscclpp::SmChannel> channels[NRANKS_PER_NODE - 1];
 
-    const int lid = threadIdx.x % WARP_SIZE;
+  const int lid = threadIdx.x % WARP_SIZE;
   if (lid < nPeer) {
     channels[lid] = smChans[lid];
   }
@@ -479,15 +478,15 @@ __global__ void __launch_bounds__(512, 1)
     __syncthreads();
 
     for (size_t idx = threadIdx.x; idx < nInt4PerChunk; idx += blockDim.x) {
-        for (int peerIdx = 0; peerIdx < NPEERS; peerIdx++) {
-            const int remoteRank = (peerIdx < rank) ? peerIdx : peerIdx + 1;
-            int4 val = scratch4Result[chunkSizePerRank * remoteRank + blockOffset + idx];
-            resultBuff4[nInt4PerRank * remoteRank + idx + offsetOfThisBlock] = val;
-        }
+      for (int peerIdx = 0; peerIdx < NPEERS; peerIdx++) {
+        const int remoteRank = (peerIdx < rank) ? peerIdx : peerIdx + 1;
+        int4 val = scratch4Result[chunkSizePerRank * remoteRank + blockOffset + idx];
+        resultBuff4[nInt4PerRank * remoteRank + idx + offsetOfThisBlock] = val;
+      }
     }
     __syncthreads();
 
-        offsetOfThisBlock += nInt4PerChunk;
+    offsetOfThisBlock += nInt4PerChunk;
     // Ensure all threads have consumed data from scratch buffer before signaling re-use in next iteration
   }
   if (restNInt4 > 0) {
@@ -527,17 +526,17 @@ __global__ void __launch_bounds__(512, 1)
     }
     __syncthreads();
     if (threadIdx.x < static_cast<uint32_t>(nPeer)) {
-        channels[threadIdx.x].signal();
-        channels[threadIdx.x].wait();
+      channels[threadIdx.x].signal();
+      channels[threadIdx.x].wait();
     }
     __syncthreads();
 
     for (size_t idx = threadIdx.x; idx < restNInt4; idx += blockDim.x) {
-        for (int peerIdx = 0; peerIdx < NPEERS; peerIdx++) {
-            const int remoteRank = (peerIdx < rank) ? peerIdx : peerIdx + 1;
-            int4 val = scratch4Result[chunkSizePerRank * remoteRank + blockOffset + idx];
-            resultBuff4[nInt4PerRank * remoteRank + idx + offsetOfThisBlock] = val;
-        }
+      for (int peerIdx = 0; peerIdx < NPEERS; peerIdx++) {
+        const int remoteRank = (peerIdx < rank) ? peerIdx : peerIdx + 1;
+        int4 val = scratch4Result[chunkSizePerRank * remoteRank + blockOffset + idx];
+        resultBuff4[nInt4PerRank * remoteRank + idx + offsetOfThisBlock] = val;
+      }
     }
     __syncthreads();
     // Ensure all threads have issued writes to outChannel
@@ -550,7 +549,6 @@ __global__ void __launch_bounds__(512, 1)
   }
 }
 
-
 template <typename T>
 __global__ void __launch_bounds__(512, 1)
     allreduce8(T* buff, T* scratch, T* resultBuff, mscclpp::DeviceHandle<mscclpp::SmChannel>* smChannels,
@@ -561,8 +559,7 @@ __global__ void __launch_bounds__(512, 1)
   // assume (nelems * sizeof(T)) is divisible by (16 * worldSize)
   const size_t nInt4 = nelems * sizeof(T) / sizeof(int4);
   size_t nInt4PerRank = nInt4 / worldSize;
-  if (nInt4 % worldSize)
-        nInt4PerRank = nInt4PerRank + 1;
+  if (nInt4 % worldSize) nInt4PerRank = nInt4PerRank + 1;
 
   auto smChans = smChannels + chanOffset;
   auto smOutChans = smOutChannels + chanOffset;
@@ -693,15 +690,14 @@ __global__ void __launch_bounds__(512, 1)
 template <typename T>
 __global__ void __launch_bounds__(512, 1)
     allreduce8Read(T* buff, T* resultBuff, mscclpp::DeviceHandle<mscclpp::SmChannel>* smChannels,
-               mscclpp::DeviceHandle<mscclpp::SmChannel>* smOutChannels, size_t channelOutDataOffset,
-               int rank, int nRanksPerNode, int worldSize, size_t nelems) {
+                   mscclpp::DeviceHandle<mscclpp::SmChannel>* smOutChannels, size_t channelOutDataOffset, int rank,
+                   int nRanksPerNode, int worldSize, size_t nelems) {
   const int nPeer = nRanksPerNode - 1;
   const size_t chanOffset = nPeer * blockIdx.x;
   // assume (nelems * sizeof(T)) is divisible by (16 * worldSize)
   const size_t nInt4 = nelems * sizeof(T) / sizeof(int4);
   size_t nInt4PerRank = nInt4 / worldSize;
-  if (nInt4 % worldSize)
-        nInt4PerRank = nInt4PerRank + 1;
+  if (nInt4 % worldSize) nInt4PerRank = nInt4PerRank + 1;
 
   auto smChans = smChannels + chanOffset;
   auto smOutChans = smOutChannels + chanOffset;
@@ -709,7 +705,7 @@ __global__ void __launch_bounds__(512, 1)
   int4* buff4 = reinterpret_cast<int4*>(buff);
   int4* resultBuff4 = reinterpret_cast<int4*>(resultBuff);
 
-   // Distribute `nInt4PerRank` across all blocks with the unit size `unitNInt4`
+  // Distribute `nInt4PerRank` across all blocks with the unit size `unitNInt4`
   constexpr size_t unitNInt4 = 512;
   const size_t maxNInt4PerBlock =
       (((nInt4PerRank + gridDim.x - 1) / gridDim.x) + unitNInt4 - 1) / unitNInt4 * unitNInt4;
@@ -746,7 +742,8 @@ __global__ void __launch_bounds__(512, 1)
     for (size_t idx = threadIdx.x; idx < nInt4PerChunk; idx += blockDim.x) {
       int4 data = buff4[nInt4PerRank * rank + idx + offsetOfThisBlock];
       for (int peerIdx = 0; peerIdx < nPeer; peerIdx++) {
-        int4 val = channels[peerIdx].read<int4>(nInt4PerRank  * rank + offsetOfThisBlock + idx);;
+        int4 val = channels[peerIdx].read<int4>(nInt4PerRank * rank + offsetOfThisBlock + idx);
+        ;
         data = add_vectors<T>(val, data);
       }
       resultBuff4[nInt4PerRank * rank + idx + offsetOfThisBlock] = data;
@@ -765,14 +762,14 @@ __global__ void __launch_bounds__(512, 1)
     if (threadIdx.x < static_cast<uint32_t>(nPeer)) {
       channels[threadIdx.x].signal();
       channels[threadIdx.x].wait();
-
     }
     __syncthreads();
 
     for (size_t idx = threadIdx.x; idx < restNInt4; idx += blockDim.x) {
       int4 data = buff4[nInt4PerRank * rank + idx + offsetOfThisBlock];
       for (int peerIdx = 0; peerIdx < nPeer; peerIdx++) {
-        int4 val = channels[peerIdx].read<int4>(nInt4PerRank  * rank + offsetOfThisBlock + idx);;
+        int4 val = channels[peerIdx].read<int4>(nInt4PerRank * rank + offsetOfThisBlock + idx);
+        ;
         data = add_vectors<T>(val, data);
       }
       resultBuff4[nInt4PerRank * rank + idx + offsetOfThisBlock] = data;
@@ -785,19 +782,18 @@ __global__ void __launch_bounds__(512, 1)
   }
 
   if (threadIdx.x < static_cast<uint32_t>(nPeer)) {
-      outChannels[threadIdx.x].signal();
-      outChannels[threadIdx.x].wait();
+    outChannels[threadIdx.x].signal();
+    outChannels[threadIdx.x].wait();
   }
   __syncthreads();
-
 }
 
 template <typename T>
 __global__ void __launch_bounds__(1024, 1)
     allreduce10(T* buff, T* scratch, T* resultBuff, mscclpp::DeviceHandle<mscclpp::SmChannel>* smChannels,
-                    mscclpp::DeviceHandle<mscclpp::SmChannel>* smScrChannels,
-               mscclpp::DeviceHandle<mscclpp::SmChannel>* smOutChannels, size_t channelOutDataOffset,
-               size_t channelScratchOffset, int rank, int nRanksPerNode, int worldSize, size_t nelems) {
+                mscclpp::DeviceHandle<mscclpp::SmChannel>* smScrChannels,
+                mscclpp::DeviceHandle<mscclpp::SmChannel>* smOutChannels, size_t channelOutDataOffset,
+                size_t channelScratchOffset, int rank, int nRanksPerNode, int worldSize, size_t nelems) {
   const int nPeer = nRanksPerNode - 1;
   const size_t chanOffset = nPeer * blockIdx.x;
   // assume (nelems * sizeof(T)) is divisible by (16 * worldSize)
@@ -821,7 +817,7 @@ __global__ void __launch_bounds__(1024, 1)
   size_t nNeededBlocks = (nInt4PerRank + maxNInt4PerBlock - 1) / maxNInt4PerBlock;
 
   constexpr size_t nInt4PerChunk = 1024 * 1024 / sizeof(int4);  // 256KB
-  int num_nodes = worldSize/NRANKS1_PER_NODE;
+  int num_nodes = worldSize / NRANKS1_PER_NODE;
 
   if (blockIdx.x >= nNeededBlocks) {
     nInt4OfThisBlock = 0;
@@ -850,42 +846,42 @@ __global__ void __launch_bounds__(1024, 1)
 
   // we can use double buffering to hide synchronization overhead
   for (size_t itr = 0; itr < nItrs; itr++) {
-    if (threadIdx.x < (NRANKS1_PER_NODE-1)) {
-        int myNode = rank/NRANKS1_PER_NODE;
-        int remote = (threadIdx.x + 1 + rank);
-        int remoteNode = remote/NRANKS1_PER_NODE;
+    if (threadIdx.x < (NRANKS1_PER_NODE - 1)) {
+      int myNode = rank / NRANKS1_PER_NODE;
+      int remote = (threadIdx.x + 1 + rank);
+      int remoteNode = remote / NRANKS1_PER_NODE;
 
-        if (remoteNode > myNode) {
-                remote = remote - NRANKS1_PER_NODE;
-        }
-        int peerIdx = remote < rank ? remote : remote - 1;
-        outChannels[peerIdx].signal();
-        outChannels[peerIdx].wait();
+      if (remoteNode > myNode) {
+        remote = remote - NRANKS1_PER_NODE;
+      }
+      int peerIdx = remote < rank ? remote : remote - 1;
+      outChannels[peerIdx].signal();
+      outChannels[peerIdx].wait();
     }
     __syncthreads();
 
-    int myNode = rank/NRANKS1_PER_NODE;
+    int myNode = rank / NRANKS1_PER_NODE;
 
-     //Reduce within an OAM
+    // Reduce within an OAM
     for (size_t idx = threadIdx.x; idx < nInt4PerChunk; idx += blockDim.x) {
       int4 data = buff4[nInt4PerRank * localRank + idx + offsetOfThisBlock];
-      for (int peerIdx = NRANKS1_PER_NODE*myNode; peerIdx < (NRANKS1_PER_NODE*myNode +
-                              NRANKS1_PER_NODE - 1); peerIdx++)      {
-        int4 val = channels[peerIdx].read<int4>(nInt4PerRank  * localRank + offsetOfThisBlock + idx);
+      for (int peerIdx = NRANKS1_PER_NODE * myNode; peerIdx < (NRANKS1_PER_NODE * myNode + NRANKS1_PER_NODE - 1);
+           peerIdx++) {
+        int4 val = channels[peerIdx].read<int4>(nInt4PerRank * localRank + offsetOfThisBlock + idx);
         data = add_vectors<T>(val, data);
       }
       scratch4[idx + blockOffset] = data;
     }
 
-    if (threadIdx.x < static_cast<uint32_t>(num_nodes-1)) {
-        int remote = (NRANKS1_PER_NODE * (threadIdx.x + 1) + rank) % worldSize;
-        int peerIdx = remote < rank ? remote : remote - 1;
-        scrChannels[peerIdx].signal();
-        scrChannels[peerIdx].wait();
+    if (threadIdx.x < static_cast<uint32_t>(num_nodes - 1)) {
+      int remote = (NRANKS1_PER_NODE * (threadIdx.x + 1) + rank) % worldSize;
+      int peerIdx = remote < rank ? remote : remote - 1;
+      scrChannels[peerIdx].signal();
+      scrChannels[peerIdx].wait();
     }
     __syncthreads();
 
-    //Reduce across OAMs
+    // Reduce across OAMs
 
     for (size_t idx = threadIdx.x; idx < nInt4PerChunk; idx += blockDim.x) {
       int4 data = scratch4[idx + blockOffset];
@@ -896,17 +892,17 @@ __global__ void __launch_bounds__(1024, 1)
         int remoteLocal = remoteRank % NRANKS1_PER_NODE;
 
         if (myLocal == remoteLocal) {
-                int4 val = scrChannels[peerIdx].read<int4>(blockOffset + idx +
-                                channelScratchOffset/sizeof(int4));
-                data = add_vectors<T>(val, data);
+          int4 val = scrChannels[peerIdx].read<int4>(blockOffset + idx + channelScratchOffset / sizeof(int4));
+          data = add_vectors<T>(val, data);
         }
       }
 
       resultBuff4[nInt4PerRank * localRank + idx + offsetOfThisBlock] = data;
 
-      for (int peerIdx = NRANKS1_PER_NODE*myNode; peerIdx < (NRANKS1_PER_NODE*myNode + NRANKS1_PER_NODE - 1); peerIdx++)      {
-        outChannels[peerIdx].write(nInt4PerRank * localRank + idx + offsetOfThisBlock +
-                        channelOutDataOffset / sizeof(int4), data);
+      for (int peerIdx = NRANKS1_PER_NODE * myNode; peerIdx < (NRANKS1_PER_NODE * myNode + NRANKS1_PER_NODE - 1);
+           peerIdx++) {
+        outChannels[peerIdx].write(
+            nInt4PerRank * localRank + idx + offsetOfThisBlock + channelOutDataOffset / sizeof(int4), data);
       }
     }
 
@@ -920,18 +916,18 @@ __global__ void __launch_bounds__(1024, 1)
   }
 
   if (restNInt4 > 0) {
-    if (threadIdx.x < (NRANKS1_PER_NODE-1)) {
-        int myNode = rank/NRANKS1_PER_NODE;
-        int remote = (threadIdx.x + 1 + rank);
-        int remoteNode = remote/NRANKS1_PER_NODE;
+    if (threadIdx.x < (NRANKS1_PER_NODE - 1)) {
+      int myNode = rank / NRANKS1_PER_NODE;
+      int remote = (threadIdx.x + 1 + rank);
+      int remoteNode = remote / NRANKS1_PER_NODE;
 
-        if (remoteNode > myNode) {
-                remote = remote - NRANKS1_PER_NODE;
-        }
-        int peerIdx = remote < rank ? remote : remote - 1;
+      if (remoteNode > myNode) {
+        remote = remote - NRANKS1_PER_NODE;
+      }
+      int peerIdx = remote < rank ? remote : remote - 1;
 
-        outChannels[peerIdx].signal();
-        outChannels[peerIdx].wait();
+      outChannels[peerIdx].signal();
+      outChannels[peerIdx].wait();
     }
     __syncthreads();
 
@@ -940,69 +936,66 @@ __global__ void __launch_bounds__(1024, 1)
       for (int peerIdx = 0; peerIdx < NPEERS; peerIdx++) {
         const int remoteRank = (peerIdx < rank) ? peerIdx : peerIdx + 1;
 
-        int myNode = rank/NRANKS1_PER_NODE;
-        int remoteNode = remoteRank/NRANKS1_PER_NODE;
+        int myNode = rank / NRANKS1_PER_NODE;
+        int remoteNode = remoteRank / NRANKS1_PER_NODE;
 
         if (myNode == remoteNode) {
-                int4 val = channels[peerIdx].read<int4>(nInt4PerRank  * localRank + offsetOfThisBlock + idx);
-                data = add_vectors<T>(val, data);
+          int4 val = channels[peerIdx].read<int4>(nInt4PerRank * localRank + offsetOfThisBlock + idx);
+          data = add_vectors<T>(val, data);
         }
       }
       scratch4[idx + blockOffset] = data;
     }
 
-    if (threadIdx.x < static_cast<uint32_t>(num_nodes-1)) {
-        int remote = (NRANKS1_PER_NODE * (threadIdx.x + 1) + rank) % worldSize;
-        int peerIdx = remote < rank ? remote : remote - 1;
-        scrChannels[peerIdx].signal();
-        scrChannels[peerIdx].wait();
+    if (threadIdx.x < static_cast<uint32_t>(num_nodes - 1)) {
+      int remote = (NRANKS1_PER_NODE * (threadIdx.x + 1) + rank) % worldSize;
+      int peerIdx = remote < rank ? remote : remote - 1;
+      scrChannels[peerIdx].signal();
+      scrChannels[peerIdx].wait();
     }
     __syncthreads();
 
     for (size_t idx = threadIdx.x; idx < restNInt4; idx += blockDim.x) {
-        int4 data = scratch4[idx + blockOffset];
-
-        for (int peerIdx = 0; peerIdx < NPEERS; peerIdx++) {
-            const int remoteRank = (peerIdx < rank) ? peerIdx : peerIdx + 1;
-            int myLocal = rank % NRANKS1_PER_NODE;
-            int remoteLocal = remoteRank % NRANKS1_PER_NODE;
-
-            if (myLocal == remoteLocal) {
-                int4 val = scrChannels[peerIdx].read<int4>(blockOffset + idx +
-                                channelScratchOffset/sizeof(int4));
-                data = add_vectors<T>(val, data);
-            }
+      int4 data = scratch4[idx + blockOffset];
+
+      for (int peerIdx = 0; peerIdx < NPEERS; peerIdx++) {
+        const int remoteRank = (peerIdx < rank) ? peerIdx : peerIdx + 1;
+        int myLocal = rank % NRANKS1_PER_NODE;
+        int remoteLocal = remoteRank % NRANKS1_PER_NODE;
+
+        if (myLocal == remoteLocal) {
+          int4 val = scrChannels[peerIdx].read<int4>(blockOffset + idx + channelScratchOffset / sizeof(int4));
+          data = add_vectors<T>(val, data);
         }
+      }
 
-       resultBuff4[nInt4PerRank * localRank + idx + offsetOfThisBlock] = data;
-	for (int peerIdx = 0; peerIdx < NPEERS; peerIdx++) {
-            const int remoteRank = (peerIdx < rank) ? peerIdx : peerIdx + 1;
-            int myNode = rank/NRANKS1_PER_NODE;
-            int remoteNode = remoteRank/NRANKS1_PER_NODE;
+      resultBuff4[nInt4PerRank * localRank + idx + offsetOfThisBlock] = data;
+      for (int peerIdx = 0; peerIdx < NPEERS; peerIdx++) {
+        const int remoteRank = (peerIdx < rank) ? peerIdx : peerIdx + 1;
+        int myNode = rank / NRANKS1_PER_NODE;
+        int remoteNode = remoteRank / NRANKS1_PER_NODE;
 
-            if (myNode == remoteNode) {
-                outChannels[peerIdx].write(nInt4PerRank * localRank + idx + offsetOfThisBlock +
-                                channelOutDataOffset / sizeof(int4), data);
-            }
+        if (myNode == remoteNode) {
+          outChannels[peerIdx].write(
+              nInt4PerRank * localRank + idx + offsetOfThisBlock + channelOutDataOffset / sizeof(int4), data);
         }
+      }
     }
     if (threadIdx.x < static_cast<uint32_t>(nPeer)) {
       outChannels[threadIdx.x].signal();
       outChannels[threadIdx.x].wait();
     }
     __syncthreads();
-
   }
-
 }
 
 template <typename T>
 cudaError_t allreduce(T* buff, T* scratch, T* resultBuff, mscclpp::DeviceHandle<mscclpp::SmChannel>* smChannels,
                       mscclpp::DeviceHandle<mscclpp::SmChannel>* smScrChannels,
-		      mscclpp::DeviceHandle<mscclpp::SmChannel>* smOutChannels, size_t channelInOffset,
+                      mscclpp::DeviceHandle<mscclpp::SmChannel>* smOutChannels, size_t channelInOffset,
                       size_t channelOutOffset, size_t channelScratchOffset, int rank, int nRanksPerNode, int worldSize,
-                      size_t nelems, cudaStream_t stream, uint32_t* deviceFlag28,
-                      uint32_t* deviceFlag56, uint32_t numScratchBuff) {
+                      size_t nelems, cudaStream_t stream, uint32_t* deviceFlag28, uint32_t* deviceFlag56,
+                      uint32_t numScratchBuff) {
   static uint32_t flag = 1;
   uint32_t* deviceFlag;
   nRanksPerNode = (worldSize < nRanksPerNode) ? worldSize : nRanksPerNode;
@@ -1010,14 +1003,15 @@ cudaError_t allreduce(T* buff, T* scratch, T* resultBuff, mscclpp::DeviceHandle<
   if (sizeof(T) * nelems < worldSize * sizeof(int)) {
     int nBlocks = nRanksPerNode - 1;
     int nThreadsPerBlock = 32;
-    allreduceAllToAll<<<nBlocks, nThreadsPerBlock, 0, stream>>>(buff, scratch, resultBuff, smChannels,
-		    channelInOffset, channelScratchOffset, rank, nRanksPerNode, worldSize, nelems, deviceFlag28, numScratchBuff);
+    allreduceAllToAll<<<nBlocks, nThreadsPerBlock, 0, stream>>>(buff, scratch, resultBuff, smChannels, channelInOffset,
+                                                                channelScratchOffset, rank, nRanksPerNode, worldSize,
+                                                                nelems, deviceFlag28, numScratchBuff);
   } else if (sizeof(T) * nelems <= (1 << 20)) {
-    int nBlocks = 4*(nRanksPerNode - 1);
+    int nBlocks = 4 * (nRanksPerNode - 1);
     int nThreadsPerBlock = 1024;
     deviceFlag = deviceFlag28;
     if (nelems >= 8192) {
-      nBlocks = 8*(nRanksPerNode - 1);
+      nBlocks = 8 * (nRanksPerNode - 1);
       nThreadsPerBlock = (nelems <= 76800) ? 512 : 1024;
       deviceFlag = deviceFlag56;
     }
@@ -1035,23 +1029,23 @@ cudaError_t allreduce(T* buff, T* scratch, T* resultBuff, mscclpp::DeviceHandle<
     int nBlocks = 8 * (nRanksPerNode - 1);
     int nThreadsPerBlock = 512;
     if (mscclppHierarchicalAllred && worldSize >= 8) {
-	nBlocks = 20;
-	allreduce10<<<nBlocks, nThreadsPerBlock, 0, stream>>>(buff, scratch, resultBuff, smChannels, smScrChannels,
-			smOutChannels, channelOutOffset, channelScratchOffset, rank, nRanksPerNode,
-			worldSize, nelems);
+      nBlocks = 20;
+      allreduce10<<<nBlocks, nThreadsPerBlock, 0, stream>>>(buff, scratch, resultBuff, smChannels, smScrChannels,
+                                                            smOutChannels, channelOutOffset, channelScratchOffset, rank,
+                                                            nRanksPerNode, worldSize, nelems);
     } else {
       if (!mscclppReadAllred) {
-	 if (mscclppDisableRemoteUbr) {
-             allreduce8Mod<<<nBlocks, nThreadsPerBlock, 0, stream>>>(buff, scratch, resultBuff, smScrChannels,
-                       channelScratchOffset, rank, nRanksPerNode, worldSize, nelems);
-	 } else {
-             allreduce8<<<nBlocks, nThreadsPerBlock, 0, stream>>>(buff, scratch, resultBuff, smScrChannels,
-		smOutChannels, channelOutOffset, channelScratchOffset, rank, nRanksPerNode,
-                   worldSize, nelems);
-	 }
+        if (mscclppDisableRemoteUbr) {
+          allreduce8Mod<<<nBlocks, nThreadsPerBlock, 0, stream>>>(
+              buff, scratch, resultBuff, smScrChannels, channelScratchOffset, rank, nRanksPerNode, worldSize, nelems);
+        } else {
+          allreduce8<<<nBlocks, nThreadsPerBlock, 0, stream>>>(buff, scratch, resultBuff, smScrChannels, smOutChannels,
+                                                               channelOutOffset, channelScratchOffset, rank,
+                                                               nRanksPerNode, worldSize, nelems);
+        }
       } else {
-	    allreduce8Read<<<nBlocks, nThreadsPerBlock, 0, stream>>>(buff, resultBuff, smChannels, smOutChannels,
-                                                         channelOutOffset, rank, nRanksPerNode,	worldSize, nelems);
+        allreduce8Read<<<nBlocks, nThreadsPerBlock, 0, stream>>>(
+            buff, resultBuff, smChannels, smOutChannels, channelOutOffset, rank, nRanksPerNode, worldSize, nelems);
       }
     }
   }
diff --git a/apps/nccl/src/nccl.cu b/apps/nccl/src/nccl.cu
index 8f66612..73e79c2 100644
--- a/apps/nccl/src/nccl.cu
+++ b/apps/nccl/src/nccl.cu
@@ -99,7 +99,7 @@ struct ncclComm {
 };
 
 struct handleInfo {
-  void * buff;
+  void* buff;
   cudaIpcMemHandle_t ipcHandle;
 };
 
@@ -236,7 +236,8 @@ static ncclResult_t ncclAllReduceFallback(const void* sendbuff, void* recvbuff,
     if (sendIt == comm->channelScratchInfos.end()) {
       std::vector<mscclpp::SmChannel> channels =
           setupSmChannels(comm, comm->remoteScratchRegMemories, const_cast<void*>((void*)sendBasePtr));
-      ChannelInfo channelInfo{channels, channels, setupSmChannelDeviceHandles(channels), setupSmChannelDeviceHandles(channels)};
+      ChannelInfo channelInfo{channels, channels, setupSmChannelDeviceHandles(channels),
+                              setupSmChannelDeviceHandles(channels)};
       sendIt = comm->channelScratchInfos.emplace(sendKey, channelInfo).first;
     }
 
@@ -253,33 +254,35 @@ static ncclResult_t ncclAllReduceFallback(const void* sendbuff, void* recvbuff,
       std::vector<mscclpp::SmChannel> channels1 =
           setupSmChannels(comm, remoteMemories1, const_cast<void*>((void*)sendBasePtr));
 
-      ChannelInfo channelInfo{channels, channels1, setupSmChannelDeviceHandles(channels), setupSmChannelDeviceHandles(channels1)};
+      ChannelInfo channelInfo{channels, channels1, setupSmChannelDeviceHandles(channels),
+                              setupSmChannelDeviceHandles(channels1)};
       sendIt = comm->channelInInfos.emplace(sendKey, channelInfo).first;
     }
 
-    if(mscclppDisableRemoteUbr == false) {
-       auto recvIt = comm->channelOutInfos.find(recvKey);
-       if (recvIt == comm->channelOutInfos.end() || mscclppDisableChannelCache == true) {
-           if (mscclppDisableChannelCache == true) {
-               recvBytes = bytes;
-               recvBasePtr = (CUdeviceptr)recvbuff;
-               offsetOut = 0;
-           }
-           remoteMemories =
-               setupRemoteMemories(comm->comm, rank, (void*)recvBasePtr, recvBytes, mscclpp::Transport::CudaIpc);
-               std::vector<mscclpp::SmChannel> outChannels =
-                   setupSmChannels(comm, remoteMemories, const_cast<void*>((void*)recvBasePtr));
-               ChannelInfo channelInfo{outChannels, outChannels, setupSmChannelDeviceHandles(outChannels), setupSmChannelDeviceHandles(outChannels)};
-               if (mscclppDisableChannelCache == true) {
-		       comm->channelInfos.push_back(channelInfo);
-                      smOutChannels = comm->channelInfos.back().smChannelDeviceHandles.get();
-               } else {
-                      recvIt = comm->channelOutInfos.emplace(recvKey, channelInfo).first;
-                      smOutChannels = recvIt->second.smChannelDeviceHandles.get();
-               }
-       } else {
-              smOutChannels = recvIt->second.smChannelDeviceHandles.get();
-       }
+    if (mscclppDisableRemoteUbr == false) {
+      auto recvIt = comm->channelOutInfos.find(recvKey);
+      if (recvIt == comm->channelOutInfos.end() || mscclppDisableChannelCache == true) {
+        if (mscclppDisableChannelCache == true) {
+          recvBytes = bytes;
+          recvBasePtr = (CUdeviceptr)recvbuff;
+          offsetOut = 0;
+        }
+        remoteMemories =
+            setupRemoteMemories(comm->comm, rank, (void*)recvBasePtr, recvBytes, mscclpp::Transport::CudaIpc);
+        std::vector<mscclpp::SmChannel> outChannels =
+            setupSmChannels(comm, remoteMemories, const_cast<void*>((void*)recvBasePtr));
+        ChannelInfo channelInfo{outChannels, outChannels, setupSmChannelDeviceHandles(outChannels),
+                                setupSmChannelDeviceHandles(outChannels)};
+        if (mscclppDisableChannelCache == true) {
+          comm->channelInfos.push_back(channelInfo);
+          smOutChannels = comm->channelInfos.back().smChannelDeviceHandles.get();
+        } else {
+          recvIt = comm->channelOutInfos.emplace(recvKey, channelInfo).first;
+          smOutChannels = recvIt->second.smChannelDeviceHandles.get();
+        }
+      } else {
+        smOutChannels = recvIt->second.smChannelDeviceHandles.get();
+      }
     }
 
     smChannels = sendIt->second.smChannelDeviceHandles1.get();
@@ -289,28 +292,31 @@ static ncclResult_t ncclAllReduceFallback(const void* sendbuff, void* recvbuff,
   switch (datatype) {
     case ncclFloat16:
       CUDACHECK(allreduce((half*)sendbuff, (half*)comm->scratchBuff.get(), (half*)recvbuff, smChannels, smScrChannels,
-	smOutChannels, offsetIn, offsetOut, offsetScratch, rank, NRANKS_PER_NODE, comm->comm->bootstrap()->getNranks(), count, stream,
-	(uint32_t*)comm->deviceFlag28.get(), (uint32_t*)comm->deviceFlag56.get(), comm->numScratchBuff));
+                          smOutChannels, offsetIn, offsetOut, offsetScratch, rank, NRANKS_PER_NODE,
+                          comm->comm->bootstrap()->getNranks(), count, stream, (uint32_t*)comm->deviceFlag28.get(),
+                          (uint32_t*)comm->deviceFlag56.get(), comm->numScratchBuff));
       break;
     case ncclFloat32:
       CUDACHECK(allreduce((float*)sendbuff, (float*)comm->scratchBuff.get(), (float*)recvbuff, smChannels,
                           smScrChannels, smOutChannels, offsetIn, offsetOut, offsetScratch,
-			  comm->comm->bootstrap()->getRank(),
-                          NRANKS_PER_NODE, comm->comm->bootstrap()->getNranks(), count, stream, (uint32_t*)comm->deviceFlag28.get(),
-			  (uint32_t*)comm->deviceFlag56.get(), comm->numScratchBuff));
+                          comm->comm->bootstrap()->getRank(), NRANKS_PER_NODE, comm->comm->bootstrap()->getNranks(),
+                          count, stream, (uint32_t*)comm->deviceFlag28.get(), (uint32_t*)comm->deviceFlag56.get(),
+                          comm->numScratchBuff));
       break;
     case ncclBfloat16:
       CUDACHECK(allreduce((__bfloat16*)sendbuff, (__bfloat16*)comm->scratchBuff.get(), (__bfloat16*)recvbuff,
                           smChannels, smScrChannels, smOutChannels, offsetIn, offsetOut, offsetScratch, rank,
-			   NRANKS_PER_NODE, comm->comm->bootstrap()->getNranks(), count, stream, (uint32_t*)comm->deviceFlag28.get(),
-			   (uint32_t*)comm->deviceFlag56.get(), comm->numScratchBuff));
+                          NRANKS_PER_NODE, comm->comm->bootstrap()->getNranks(), count, stream,
+                          (uint32_t*)comm->deviceFlag28.get(), (uint32_t*)comm->deviceFlag56.get(),
+                          comm->numScratchBuff));
       break;
     case ncclInt32:
     case ncclUint32:
       CUDACHECK(allreduce((int*)sendbuff, (int*)comm->scratchBuff.get(), (int*)recvbuff, smChannels, smScrChannels,
-			  smOutChannels, offsetIn, offsetOut, offsetScratch, comm->comm->bootstrap()->getRank(),
-			  NRANKS_PER_NODE, comm->comm->bootstrap()->getNranks(), count, stream, (uint32_t*)comm->deviceFlag28.get(),
-			  (uint32_t*)comm->deviceFlag56.get(), comm->numScratchBuff));
+                          smOutChannels, offsetIn, offsetOut, offsetScratch, comm->comm->bootstrap()->getRank(),
+                          NRANKS_PER_NODE, comm->comm->bootstrap()->getNranks(), count, stream,
+                          (uint32_t*)comm->deviceFlag28.get(), (uint32_t*)comm->deviceFlag56.get(),
+                          comm->numScratchBuff));
       break;
     default:
       WARN("datatype is invalid");
@@ -349,9 +355,9 @@ static ncclResult_t ncclAllGatherFallback(const void* sendbuff, void* recvbuff,
   auto it = comm->channelOutInfos.find(recvKey);
   if (it == comm->channelOutInfos.end() || mscclppDisableChannelCache == true) {
     if (mscclppDisableChannelCache == true) {
-        recvBytes = bytes;
-        recvBasePtr = (CUdeviceptr)recvbuff;
-        offsetOut = 0;
+      recvBytes = bytes;
+      recvBasePtr = (CUdeviceptr)recvbuff;
+      offsetOut = 0;
     }
     std::vector<mscclpp::RegisteredMemory> remoteMemories = setupRemoteMemories(
         comm->comm, rank, const_cast<void*>((void*)recvBasePtr), recvBytes, mscclpp::Transport::CudaIpc);
@@ -360,16 +366,17 @@ static ncclResult_t ncclAllGatherFallback(const void* sendbuff, void* recvbuff,
     std::vector<mscclpp::DeviceHandle<mscclpp::SmChannel>> smChannelDeviceHandles;
     std::transform(channels.begin(), channels.end(), std::back_inserter(smChannelDeviceHandles),
                    [](const mscclpp::SmChannel& smChannel) { return mscclpp::deviceHandle(smChannel); });
-    ChannelInfo channelInfo{channels, channels, setupSmChannelDeviceHandles(channels), setupSmChannelDeviceHandles(channels)};
+    ChannelInfo channelInfo{channels, channels, setupSmChannelDeviceHandles(channels),
+                            setupSmChannelDeviceHandles(channels)};
     if (mscclppDisableChannelCache == true) {
-	comm->channelInfos.push_back(channelInfo);
-	smChannels = comm->channelInfos.back().smChannelDeviceHandles.get();
+      comm->channelInfos.push_back(channelInfo);
+      smChannels = comm->channelInfos.back().smChannelDeviceHandles.get();
     } else {
-       it = comm->channelOutInfos.emplace(recvKey, channelInfo).first;
-       smChannels = it->second.smChannelDeviceHandles.get();
+      it = comm->channelOutInfos.emplace(recvKey, channelInfo).first;
+      smChannels = it->second.smChannelDeviceHandles.get();
     }
   } else {
-	smChannels = it->second.smChannelDeviceHandles.get();
+    smChannels = it->second.smChannelDeviceHandles.get();
   }
 
   if ((char*)sendbuff == (char*)recvbuff + rank * sendcount) {
@@ -628,50 +635,51 @@ NCCL_API ncclResult_t ncclCommRegister(ncclComm_t comm, void* buff, size_t size,
   // Creating the channels
   auto buffIt = comm->channelScratchInfos.find(buffKey);
   if (buffIt == comm->channelScratchInfos.end()) {
-     std::vector<mscclpp::SmChannel> channels =
-          setupSmChannels(comm, comm->remoteScratchRegMemories, const_cast<void*>((void*)buffBasePtr));
-     ChannelInfo channelInfo{channels, channels, setupSmChannelDeviceHandles(channels), setupSmChannelDeviceHandles(channels)};
-     buffIt = comm->channelScratchInfos.emplace(buffKey, channelInfo).first;
+    std::vector<mscclpp::SmChannel> channels =
+        setupSmChannels(comm, comm->remoteScratchRegMemories, const_cast<void*>((void*)buffBasePtr));
+    ChannelInfo channelInfo{channels, channels, setupSmChannelDeviceHandles(channels),
+                            setupSmChannelDeviceHandles(channels)};
+    buffIt = comm->channelScratchInfos.emplace(buffKey, channelInfo).first;
   }
   auto sendIt = comm->channelInInfos.find(buffKey);
   if (sendIt == comm->channelInInfos.end()) {
-      std::vector<mscclpp::SmChannel> channels =
-          setupSmChannels(comm, comm->remoteScratchRegMemories, const_cast<void*>((void*)buffBasePtr));
+    std::vector<mscclpp::SmChannel> channels =
+        setupSmChannels(comm, comm->remoteScratchRegMemories, const_cast<void*>((void*)buffBasePtr));
 
-      remoteMemories =
-          setupRemoteMemories(comm->comm, rank, (void*)buffBasePtr, buffBytes, mscclpp::Transport::CudaIpc);
-      std::vector<mscclpp::SmChannel> channels1 =
-          setupSmChannels(comm, remoteMemories, const_cast<void*>((void*)buffBasePtr));
+    remoteMemories = setupRemoteMemories(comm->comm, rank, (void*)buffBasePtr, buffBytes, mscclpp::Transport::CudaIpc);
+    std::vector<mscclpp::SmChannel> channels1 =
+        setupSmChannels(comm, remoteMemories, const_cast<void*>((void*)buffBasePtr));
 
-      ChannelInfo channelInfo{channels, channels1, setupSmChannelDeviceHandles(channels), setupSmChannelDeviceHandles(channels1)};
-      sendIt = comm->channelInInfos.emplace(buffKey, channelInfo).first;
+    ChannelInfo channelInfo{channels, channels1, setupSmChannelDeviceHandles(channels),
+                            setupSmChannelDeviceHandles(channels1)};
+    sendIt = comm->channelInInfos.emplace(buffKey, channelInfo).first;
   }
   auto recvIt = comm->channelOutInfos.find(buffKey);
-    if (recvIt == comm->channelOutInfos.end()) {
-      remoteMemories =
-          setupRemoteMemories(comm->comm, rank, (void*)buffBasePtr, buffBytes, mscclpp::Transport::CudaIpc);
-      std::vector<mscclpp::SmChannel> outChannels =
-          setupSmChannels(comm, remoteMemories, const_cast<void*>((void*)buffBasePtr));
-      ChannelInfo channelInfo{outChannels, outChannels, setupSmChannelDeviceHandles(outChannels), setupSmChannelDeviceHandles(outChannels)};
-      recvIt = comm->channelOutInfos.emplace(buffKey, channelInfo).first;
+  if (recvIt == comm->channelOutInfos.end()) {
+    remoteMemories = setupRemoteMemories(comm->comm, rank, (void*)buffBasePtr, buffBytes, mscclpp::Transport::CudaIpc);
+    std::vector<mscclpp::SmChannel> outChannels =
+        setupSmChannels(comm, remoteMemories, const_cast<void*>((void*)buffBasePtr));
+    ChannelInfo channelInfo{outChannels, outChannels, setupSmChannelDeviceHandles(outChannels),
+                            setupSmChannelDeviceHandles(outChannels)};
+    recvIt = comm->channelOutInfos.emplace(buffKey, channelInfo).first;
   }
 
   cudaIpcMemHandle_t ipcHandle;
   MSCCLPP_CUDATHROW(cudaIpcGetMemHandle(&ipcHandle, buffBasePtr));
 
-  struct handleInfo *p = (struct handleInfo *) malloc(sizeof(struct handleInfo));
+  struct handleInfo* p = (struct handleInfo*)malloc(sizeof(struct handleInfo));
   p->buff = buffBasePtr;
   p->ipcHandle = ipcHandle;
   *handle = p;
 
   auto regIt = comm->regHandles.find(buffKey);
   if (regIt == comm->regHandles.end()) {
-        comm->regHandles[buffKey] = ipcHandle;
+    comm->regHandles[buffKey] = ipcHandle;
   }
 
   auto it = comm->handleKeys.find(*handle);
   if (it == comm->handleKeys.end()) {
-     comm->handleKeys[*handle] = buffKey;
+    comm->handleKeys[*handle] = buffKey;
   }
 
   return ncclSuccess;
@@ -679,43 +687,41 @@ NCCL_API ncclResult_t ncclCommRegister(ncclComm_t comm, void* buff, size_t size,
 
 NCCL_API ncclResult_t ncclCommDeregister(ncclComm_t comm, void* handle) {
   if (comm && handle) {
-     channelKey buffKey = comm->handleKeys[handle];
-
-     auto scratchIt = comm->channelScratchInfos.find(buffKey);
-     if (scratchIt != comm->channelScratchInfos.end()) {
-        comm->channelScratchInfos.erase(scratchIt);
-     }
-
-     auto inIt = comm->channelInInfos.find(buffKey);
-     if (inIt != comm->channelInInfos.end()) {
-        comm->channelInInfos.erase(inIt);
-     }
-
-     auto outIt = comm->channelOutInfos.find(buffKey);
-     if (outIt != comm->channelOutInfos.end()) {
-        comm->channelOutInfos.erase(outIt);
-     }
-     comm->regHandles.erase(buffKey);
-     comm->handleKeys.erase(handle);
-     free(handle);
+    channelKey buffKey = comm->handleKeys[handle];
+
+    auto scratchIt = comm->channelScratchInfos.find(buffKey);
+    if (scratchIt != comm->channelScratchInfos.end()) {
+      comm->channelScratchInfos.erase(scratchIt);
+    }
+
+    auto inIt = comm->channelInInfos.find(buffKey);
+    if (inIt != comm->channelInInfos.end()) {
+      comm->channelInInfos.erase(inIt);
+    }
+
+    auto outIt = comm->channelOutInfos.find(buffKey);
+    if (outIt != comm->channelOutInfos.end()) {
+      comm->channelOutInfos.erase(outIt);
+    }
+    comm->regHandles.erase(buffKey);
+    comm->handleKeys.erase(handle);
+    free(handle);
   }
   return ncclSuccess;
 }
 
-bool mscclpp_BuffIsRegistered(ncclComm_t comm, const void* buff){
-  if(buff == nullptr)
-    return false;
+bool mscclpp_BuffIsRegistered(ncclComm_t comm, const void* buff) {
+  if (buff == nullptr) return false;
   size_t buffBytes;
   CUdeviceptr buffBasePtr;
   MSCCLPP_CUTHROW(cuMemGetAddressRange(&buffBasePtr, &buffBytes, (CUdeviceptr)buff));
   channelKey buffKey{(void*)buffBasePtr, buffBytes};
   auto buffIt = comm->regHandles.find(buffKey);
-  bool registered =  buffIt != comm->regHandles.end();
+  bool registered = buffIt != comm->regHandles.end();
   return registered;
 }
-size_t
-mscclpp_BufferSize(ncclComm_t comm, void* handle){
-  if (!(comm && handle)){
+size_t mscclpp_BufferSize(ncclComm_t comm, void* handle) {
+  if (!(comm && handle)) {
     return 0;
   }
   auto buffKeyIt = comm->handleKeys.find(handle);
@@ -767,7 +773,8 @@ NCCL_API ncclResult_t ncclBroadcastFallback(const void* sendbuff, void* recvbuff
     std::vector<mscclpp::DeviceHandle<mscclpp::SmChannel>> smChannelDeviceHandles;
     std::transform(channels.begin(), channels.end(), std::back_inserter(smChannelDeviceHandles),
                    [](const mscclpp::SmChannel& smChannel) { return mscclpp::deviceHandle(smChannel); });
-    ChannelInfo channelInfo{channels, channels, setupSmChannelDeviceHandles(channels), setupSmChannelDeviceHandles(channels)};
+    ChannelInfo channelInfo{channels, channels, setupSmChannelDeviceHandles(channels),
+                            setupSmChannelDeviceHandles(channels)};
     it = comm->channelOutInfos.emplace(recvKey, channelInfo).first;
   }
 
diff --git a/src/debug.cc b/src/debug.cc
index d49eb4a..c2db8d6 100644
--- a/src/debug.cc
+++ b/src/debug.cc
@@ -58,38 +58,38 @@ void mscclppDebugInit() {
 
   const char* disable_channel_cache = getenv("MSCCLPP_DISABLE_CHANNEL_CACHE");
   if (disable_channel_cache == NULL) {
-     mscclppDisableChannelCache = false;
+    mscclppDisableChannelCache = false;
   } else if (strcasecmp(disable_channel_cache, "TRUE") == 0) {
-     mscclppDisableChannelCache = true;
+    mscclppDisableChannelCache = true;
   } else {
-     mscclppDisableChannelCache = false;
+    mscclppDisableChannelCache = false;
   }
 
   const char* enable_read_allred = getenv("MSCCLPP_READ_ALLRED");
   if (enable_read_allred == NULL) {
-     mscclppReadAllred = false;
+    mscclppReadAllred = false;
   } else if (strcasecmp(enable_read_allred, "TRUE") == 0) {
-     mscclppReadAllred = true;
+    mscclppReadAllred = true;
   } else {
-     mscclppReadAllred = false;
+    mscclppReadAllred = false;
   }
 
   const char* enable_hierarchical_allred = getenv("MSCCLPP_HIERARCHICAL_ALLRED");
   if (enable_hierarchical_allred == NULL) {
-     mscclppHierarchicalAllred = false;
+    mscclppHierarchicalAllred = false;
   } else if (strcasecmp(enable_hierarchical_allred, "TRUE") == 0) {
-     mscclppHierarchicalAllred = true;
+    mscclppHierarchicalAllred = true;
   } else {
-     mscclppHierarchicalAllred = false;
+    mscclppHierarchicalAllred = false;
   }
 
-  const char*  disable_remote_ubr = getenv("MSCCLPP_DISABLE_REMOTE_UBR");
+  const char* disable_remote_ubr = getenv("MSCCLPP_DISABLE_REMOTE_UBR");
   if (disable_remote_ubr == NULL) {
-     mscclppDisableRemoteUbr = false;
+    mscclppDisableRemoteUbr = false;
   } else if (strcasecmp(disable_remote_ubr, "TRUE") == 0) {
-     mscclppDisableRemoteUbr = true;
+    mscclppDisableRemoteUbr = true;
   } else {
-     mscclppDisableRemoteUbr = false;
+    mscclppDisableRemoteUbr = false;
   }
 
   /* Parse the MSCCLPP_DEBUG_SUBSYS env var
diff --git a/src/ib.cc b/src/ib.cc
index 2269923..79b906b 100644
--- a/src/ib.cc
+++ b/src/ib.cc
@@ -56,9 +56,9 @@ IbMr::IbMr(ibv_pd* pd, void* buff, std::size_t size) : buff(buff) {
                                   IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_REMOTE_READ |
                                       IBV_ACCESS_RELAXED_ORDERING | IBV_ACCESS_REMOTE_ATOMIC);
 #else
-  this->mr = IBVerbs::ibv_reg_mr2(pd, reinterpret_cast<void*>(addr), pages * pageSize,
-                                  IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_REMOTE_READ |
-                                      IBV_ACCESS_REMOTE_ATOMIC);
+  this->mr = IBVerbs::ibv_reg_mr2(
+      pd, reinterpret_cast<void*>(addr), pages * pageSize,
+      IBV_ACCESS_LOCAL_WRITE | IBV_ACCESS_REMOTE_WRITE | IBV_ACCESS_REMOTE_READ | IBV_ACCESS_REMOTE_ATOMIC);
 #endif
 
   if (this->mr == nullptr) {
