diff --git a/src/numa.cc b/src/numa.cc
index d72c99e..16c903d 100644
--- a/src/numa.cc
+++ b/src/numa.cc
@@ -26,6 +26,7 @@ namespace mscclpp {
 
 MSCCLPP_API_CPP int getDeviceNumaNode(int cudaDev) {
   std::string busId = getBusId(cudaDev);
+  busId[busId.length() - 1] = '0';
   std::string file_str = "/sys/bus/pci/devices/" + busId + "/numa_node";
   std::ifstream file(file_str);
   int numaNode;
