From 40f56160d3a6eab7279f7e9fbd2e5b6958c6d9b3 Mon Sep 17 00:00:00 2001
From: Thomas Huber <thomas.huber@amd.com>
Date: Fri, 17 Oct 2025 05:23:46 +0000
Subject: [PATCH] Disable external MPI requirement for RCCL build integration

---
 CMakeLists.txt | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9bd60d0..edf8ea7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -151,9 +151,9 @@ if (NOT BUILD_TESTS_ONLY)
   #############################################################################
   # PACKAGE DEPENDENCIES
   #############################################################################
-  find_package(MPI)
+#find_package(MPI)
   if (MPI_FOUND)
-    set(HAVE_EXTERNAL_MPI ON)
+    set(HAVE_EXTERNAL_MPI OFF)
   else()
     set(HAVE_EXTERNAL_MPI OFF)
     set(BUILD_FUNCTIONAL_TESTS OFF)
-- 
2.34.1

