diff --git a/include/mscclpp/env.hpp b/include/mscclpp/env.hpp
index 6708628..3f96b31 100644
--- a/include/mscclpp/env.hpp
+++ b/include/mscclpp/env.hpp
@@ -30,6 +30,7 @@ class Env {
   const std::string executionPlanDir;
   const std::string npkitDumpDir;
   const bool cudaIpcUseDefaultStream;
+  const int ibGidIndex;
 
  private:
   Env();
diff --git a/src/bootstrap/socket.cc b/src/bootstrap/socket.cc
index 8c012c2..5dc43b6 100644
--- a/src/bootstrap/socket.cc
+++ b/src/bootstrap/socket.cc
@@ -327,6 +327,8 @@ int FindInterfaces(char* ifNames, union SocketAddress* ifAddrs, int ifNameMaxSiz
         nIfs = FindInterfaceMatchSubnet(ifNames, ifAddrs, &idAddr, ifNameMaxSize, maxIfs);
       }
     }
+    // Look for Ethernet if no IB interface was found
+    if (nIfs == 0) nIfs = findInterfaces("en", ifNames, ifAddrs, sock_family, ifNameMaxSize, maxIfs);
     // Then look for anything else (but not docker or lo)
     if (nIfs == 0) nIfs = findInterfaces("^docker,lo", ifNames, ifAddrs, sock_family, ifNameMaxSize, maxIfs);
     // Finally look for docker, then lo.
diff --git a/src/env.cpp b/src/env.cpp
index 625de0a..bcb847e 100644
--- a/src/env.cpp
+++ b/src/env.cpp
@@ -19,8 +19,9 @@ T readEnv(const std::string &envName, const T &defaultValue) {
     return atoi(envCstr);
   } else if constexpr (std::is_same_v<T, bool>) {
     return (std::string(envCstr) != "0");
+  } else {
+    return T(envCstr);
   }
-  return T(envCstr);
 }
 
 template <typename T>
@@ -61,7 +62,8 @@ Env::Env()
       commId(readEnv<std::string>("MSCCLPP_COMM_ID", "")),
       executionPlanDir(readEnv<std::string>("MSCCLPP_EXECUTION_PLAN_DIR", "")),
       npkitDumpDir(readEnv<std::string>("MSCCLPP_NPKIT_DUMP_DIR", "")),
-      cudaIpcUseDefaultStream(readEnv<bool>("MSCCLPP_CUDAIPC_USE_DEFAULT_STREAM", false)) {}
+      cudaIpcUseDefaultStream(readEnv<bool>("MSCCLPP_CUDAIPC_USE_DEFAULT_STREAM", false)),
+      ibGidIndex(readEnv<int>("MSCCLPP_IB_GID_INDEX", -1))  {}
 
 std::shared_ptr<Env> env() {
   static std::shared_ptr<Env> globalEnv = std::shared_ptr<Env>(new Env());
@@ -80,6 +82,7 @@ std::shared_ptr<Env> env() {
     logEnv("MSCCLPP_EXECUTION_PLAN_DIR", globalEnv->executionPlanDir);
     logEnv("MSCCLPP_NPKIT_DUMP_DIR", globalEnv->npkitDumpDir);
     logEnv("MSCCLPP_CUDAIPC_USE_DEFAULT_STREAM", globalEnv->cudaIpcUseDefaultStream);
+    logEnv("MSCCLPP_IB_GID_INDEX", globalEnv->ibGidIndex);
   }
   return globalEnv;
 }
diff --git a/src/ib.cc b/src/ib.cc
index 542527e..7cddc55 100644
--- a/src/ib.cc
+++ b/src/ib.cc
@@ -20,6 +20,8 @@
 #include "debug.h"
 #if defined(USE_IBVERBS)
 #include "ibverbs_wrapper.hpp"
+#include <arpa/inet.h>
+#include <fcntl.h>
 #endif  // defined(USE_IBVERBS)
 
 #if !defined(__HIP_PLATFORM_AMD__)
@@ -40,6 +42,134 @@ namespace mscclpp {
 
 #if defined(USE_IBVERBS)
 
+static std::vector<std::string> getActiveIbDeviceNames(int& numActiveDevices) {
+  int count;
+  std::vector<std::string> activeDevices;
+  struct ibv_device** devices = IBVerbs::ibv_get_device_list(&count);
+  if(!devices) {
+    numActiveDevices = 0;
+    return activeDevices;
+  }
+  for (int i = 0; i < count; ++i) {
+    IbCtx ctx(devices[i]->name);
+    if(ctx.getAnyActivePort() < 0) continue;
+    activeDevices.push_back(devices[i]->name);
+  }
+  numActiveDevices = activeDevices.size();
+  IBVerbs::ibv_free_device_list(devices);
+  return activeDevices;
+}
+
+static bool isConfiguredGid(union ibv_gid* gid)
+{
+  const struct in6_addr *a = (struct in6_addr *)gid->raw;
+  int trailer = (a->s6_addr32[1] | a->s6_addr32[2] | a->s6_addr32[3]);
+  if (((a->s6_addr32[0] | trailer) == 0UL) ||
+      ((a->s6_addr32[0] == htonl(0xfe800000)) && (trailer == 0UL))) {
+    return false;
+  }
+  return true;
+}
+
+static bool isLinkLocalGid(union ibv_gid const& gid)
+{
+  const struct in6_addr *a = (struct in6_addr *) gid.raw;
+  if (a->s6_addr32[0] == htonl(0xfe800000) && a->s6_addr32[1] == 0UL) {
+    return true;
+  }
+  return false;
+}
+
+static int getRoceVersionNumber(struct ibv_context* const& context, int const& portNum, int const& gidIndex)
+{
+  char const* deviceName = ibv_get_device_name(context->device);
+  char gidRoceVerStr[16]      = {};
+  char roceTypePath[PATH_MAX] = {};
+
+  sprintf(roceTypePath, "/sys/class/infiniband/%s/ports/%d/gid_attrs/types/%d",
+          deviceName, portNum, gidIndex);
+
+
+  int fd = open(roceTypePath, O_RDONLY);
+  if (fd == -1) {
+    std::stringstream err;
+    err << "Failed while opening RoCE file path " << roceTypePath;
+    throw mscclpp::Error(err.str(), ErrorCode::InternalError);
+  }
+
+  int ret = read(fd, gidRoceVerStr, 15);
+  close(fd);
+
+  if (ret == -1) {
+    std::stringstream err;
+    err << "Failed while reading RoCE version";
+    throw mscclpp::Error(err.str(), ErrorCode::InternalError);
+  }
+
+  if (strlen(gidRoceVerStr)) {
+    if (strncmp(gidRoceVerStr, "IB/RoCE v1", strlen("IB/RoCE v1")) == 0
+        || strncmp(gidRoceVerStr, "RoCE v1", strlen("RoCE v1")) == 0) {
+      return 1;
+    }
+    else if (strncmp(gidRoceVerStr, "RoCE v2", strlen("RoCE v2")) == 0) {
+      return 2;
+    }
+  }
+  return -1;
+}
+
+static bool isIPv4MappedIPv6(const union ibv_gid &gid)
+{
+  // look for ::ffff:x.x.x.x format
+  // From Broadcom documentation
+  // https://techdocs.broadcom.com/us/en/storage-and-ethernet-connectivity/ethernet-nic-controllers/bcm957xxx/adapters/frequently-asked-questions1.html
+  // "The IPv4 address is really an IPv4 address mapped into the IPv6 address space.
+  // This can be identified by 80 “0” bits, followed by 16 “1” bits (“FFFF” in hexadecimal)
+  // followed by the original 32-bit IPv4 address."
+  return (gid.global.subnet_prefix == 0    &&
+          gid.raw[8]               == 0    &&
+          gid.raw[9]               == 0    &&
+          gid.raw[10]              == 0xff &&
+          gid.raw[11]              == 0xff);
+}
+
+
+static uint8_t getGidIndex(struct ibv_context* context, int const& ibPort, int const& gidTblLen)
+{
+  union ibv_gid gid;
+  GidPriority highestPriority = GidPriority::UNKNOWN;
+  int gidIndex = -1;
+  for (int i = 0; i < gidTblLen; ++i) {
+    if (ibv_query_gid(context, ibPort , i, &gid) != 0) continue;
+    if (!isConfiguredGid(&gid)) continue;
+    int gidCurrRoceVersion;
+    gidCurrRoceVersion = getRoceVersionNumber(context, ibPort, i);
+    if(gidCurrRoceVersion < 1) continue;
+    GidPriority currPriority;
+    if (isIPv4MappedIPv6(gid)) {
+      currPriority = (gidCurrRoceVersion == 2) ? GidPriority::ROCEV2_IPV4 : GidPriority::ROCEV1_IPV4;
+    } else if (!isLinkLocalGid(gid)) {
+      currPriority = (gidCurrRoceVersion == 2) ? GidPriority::ROCEV2_IPV6 : GidPriority::ROCEV1_IPV6;
+    } else {
+      currPriority = (gidCurrRoceVersion == 2) ? GidPriority::ROCEV2_LINK_LOCAL : GidPriority::ROCEV1_LINK_LOCAL;
+    }
+    if(currPriority > highestPriority) {
+      highestPriority = currPriority;
+      gidIndex = i;
+    }
+  }
+
+  if(highestPriority == GidPriority::UNKNOWN){
+    std::stringstream err;
+    err << "Auto GetGidIndex failed. Try setting MSCCLPP_IB_GID_INDEX directly";
+    throw mscclpp::Error(err.str(), ErrorCode::SystemError);
+  }
+  if(gidIndex >= UINT8_MAX || gidIndex < 0) {
+    throw mscclpp::Error("Invalid auto-deteced gidIndex : " + std::to_string(gidIndex), ErrorCode::SystemError);
+  }
+  return static_cast<uint8_t>(gidIndex);
+}
+
 IbMr::IbMr(ibv_pd* pd, void* buff, std::size_t size) : buff(buff) {
   if (size == 0) {
     throw std::invalid_argument("invalid size: " + std::to_string(size));
@@ -117,8 +247,9 @@ IbQp::IbQp(ibv_context* ctx, ibv_pd* pd, int port, int maxCqSize, int maxCqPollN
   this->info.is_grh = (portAttr.flags & IBV_QPF_GRH_REQUIRED);
 
   if (portAttr.link_layer != IBV_LINK_LAYER_INFINIBAND || this->info.is_grh) {
+    this->info.gidIndex = (env()->ibGidIndex < 0) ? getGidIndex(ctx, port, portAttr.gid_tbl_len) : env()->ibGidIndex;
     union ibv_gid gid;
-    if (IBVerbs::ibv_query_gid(ctx, port, 0, &gid) != 0) {
+    if (IBVerbs::ibv_query_gid(ctx, port, this->info.gidIndex, &gid) != 0) {
       std::stringstream err;
       err << "ibv_query_gid failed (errno " << errno << ")";
       throw mscclpp::IbError(err.str(), errno);
@@ -164,7 +295,7 @@ void IbQp::rtr(const IbQpInfo& info) {
     qp_attr.ah_attr.grh.dgid.global.subnet_prefix = info.spn;
     qp_attr.ah_attr.grh.dgid.global.interface_id = info.iid;
     qp_attr.ah_attr.grh.flow_label = 0;
-    qp_attr.ah_attr.grh.sgid_index = 0;
+    qp_attr.ah_attr.grh.sgid_index = this->info.gidIndex;
     qp_attr.ah_attr.grh.hop_limit = 255;
     qp_attr.ah_attr.grh.traffic_class = 0;
   } else {
@@ -385,7 +516,7 @@ const IbMr* IbCtx::registerMr(void* buff, std::size_t size) {
 
 MSCCLPP_API_CPP int getIBDeviceCount() {
   int num;
-  IBVerbs::ibv_get_device_list(&num);
+  auto const& dev = getActiveIbDeviceNames(num);
   return num;
 }
 
@@ -442,20 +573,20 @@ MSCCLPP_API_CPP std::string getIBDeviceName(Transport ibTransport) {
   }
 
   int num;
-  struct ibv_device** devices = IBVerbs::ibv_get_device_list(&num);
+  auto const& devices = getActiveIbDeviceNames(num);
   if (ibTransportIndex >= num) {
     std::stringstream ss;
     ss << "IB transport out of range: " << ibTransportIndex << " >= " << num;
     throw Error(ss.str(), ErrorCode::InvalidUsage);
   }
-  return devices[ibTransportIndex]->name;
+  return devices[ibTransportIndex];
 }
 
 MSCCLPP_API_CPP Transport getIBTransportByDeviceName(const std::string& ibDeviceName) {
   int num;
-  struct ibv_device** devices = IBVerbs::ibv_get_device_list(&num);
+  auto const& devices = getActiveIbDeviceNames(num);
   for (int i = 0; i < num; ++i) {
-    if (ibDeviceName == devices[i]->name) {
+    if (ibDeviceName == devices[i]) {
       switch (i) {  // TODO: get rid of this ugly switch
         case 0:
           return Transport::IB0;
diff --git a/src/include/ib.hpp b/src/include/ib.hpp
index 4d85299..3018d54 100644
--- a/src/include/ib.hpp
+++ b/src/include/ib.hpp
@@ -54,6 +54,7 @@ struct IbQpInfo {
   int mtu;
   uint64_t iid;
   bool is_grh;
+  uint8_t gidIndex;
 };
 
 enum class WsStatus {
@@ -132,9 +133,9 @@ class IbCtx {
 
   const std::string& getDevName() const { return this->devName; };
 
+  int getAnyActivePort() const;
  private:
   bool isPortUsable(int port) const;
-  int getAnyActivePort() const;
 
   const std::string devName;
   ibv_context* ctx;
@@ -142,7 +143,21 @@ class IbCtx {
   std::list<std::unique_ptr<IbQp>> qps;
   std::list<std::unique_ptr<IbMr>> mrs;
 };
-
+/**
+  * Enumeration of GID priority
+  *
+  * @note These are the GID types ordered in priority from lowest (0) to highest
+  */
+enum GidPriority
+{
+  UNKNOWN           = -1,                      ///< Default
+  ROCEV1_LINK_LOCAL = 0,                       ///< RoCEv1 Link-local
+  ROCEV2_LINK_LOCAL = 1,                       ///< RoCEv2 Link-local fe80::/10
+  ROCEV1_IPV6       = 2,                       ///< RoCEv1 IPv6
+  ROCEV2_IPV6       = 3,                       ///< RoCEv2 IPv6
+  ROCEV1_IPV4       = 4,                       ///< RoCEv1 IPv4-mapped IPv6
+  ROCEV2_IPV4       = 5,                       ///< RoCEv2 IPv4-mapped IPv6 ::ffff:192.168.x.x
+};
 }  // namespace mscclpp
 
 #endif  // MSCCLPP_IB_HPP_
